Feature/move load steer on vaps from beerocks controller conf to uci

adam1985d opened the pull request:
Currently, the load_steer_on_vaps that is used by the optimal-path-task
is read from the beerocks_controller.conf and is constructed differently
for each platform according to compilation flags.
    
Added the hostap_iface_steer_vaps param to each radio in the
prplmesh config file. The load_steer_on_vaps will be constructed from
the hostap_iface_steer_vaps of all radios read from UCI.

After moving the 'load_steer_on_vaps' to the UCI, the 'load_steer_on_vaps' parameter and the implementation of reading it from beerock_controller.conf is removed.

The compilation flags mentioned above are not used elsewhere in the code and are also removed:
```
BEEROCKS_WLAN1_STEERING_VAPS
BEEROCKS_WLAN2_STEERING_VAPS
BEEROCKS_WLAN3_STEERING_VAPS
```

kantera800 commented on framework/platform/bpl/include/bpl/bpl_cfg.h:
```diff
@@ -62,8 +62,9 @@ namespace bpl {
-#define BPL_SERIAL_NUMBER_LEN (64 + 1) /* Maximal length of SERIAL_NUMBER */
+#define BPL_USER_PASS_LEN (64 + 1)           /* Maximal length of USER_PASS */
+#define BPL_SERIAL_NUMBER_LEN (64 + 1)       /* Maximal length of SERIAL_NUMBER */
+#define BPL_LOAD_STEER_ON_VAPS_LEN (127 + 1) /* Maximal length of LOAD_STEER_ON_VAPS string */
```
Just curious why  127 characters? 64 is quite enough even for 8 vaps

adam1985d replied:
Currently, the VAP names are `wlan_._` - but:
1. there is a plan to remove the dependency of VAP names to the hostap_iface name.
2. if for some reason the hostap_iface names are changed to something longer (even if we still use `._` for VAPs) we may reach the 64 characters limit very easily for 8 VAPs

kantera800 commented on framework/platform/bpl/linux/bpl_cfg.cpp:
```diff
@@ -162,6 +162,46 @@ int cfg_get_certification_mode()
+                    load_steer_on_vaps_str.append(",");
+                }
+                // for linux implementation the wlan?.0 vaps are used for band steering
+                load_steer_on_vaps_str.append(std::string(ifname) + ".0");
```
Why the special treatment, dummy mode doesn't have UCI configurations?

adam1985d replied:
no - it doesn't
it has a `platform_prplmesh_db` file but you can see right away it covers minimal configurations and doesn't support the full configuration provided by the prplmesh config file

kantera800 reviewed (approved):
Good job. 
Please mention any platform testing that you did since we don't have any ci for optimal path yet. 
Also, you should open a task to move UCI db configurations to rdkb feed.

adam1985d commented:
> Good job.
> Please mention any platform testing that you did since we don't have any ci for optimal path yet.
> Also, you should open a task to move UCI db configurations to rdkb feed.

I will add a test description before I change to a non-draft-PR, along with a more elaborate description of the variation between the platforms.

regardings the second part - let's have a short session with @vitalybu @morantr and @tomereli , as I understood from @morantr this implementation was intentional.

tomereli commented on framework/platform/bpl/linux/bpl_cfg.cpp:
```diff
@@ -162,6 +162,13 @@ int cfg_get_certification_mode()
+int cfg_get_load_steer_on_vaps(int num_of_interfaces,
+                               char load_steer_on_vaps[BPL_LOAD_STEER_ON_VAPS_LEN])
+{
+
```
remove blank line

adam1985d replied:
was removed in the implementation commit of that function

tomereli commented on framework/platform/bpl/uci/cfg/bpl_cfg.cpp:
```diff
@@ -110,6 +110,13 @@ int cfg_get_certification_mode()
+int cfg_get_load_steer_on_vaps(int num_of_interfaces,
+                               char load_steer_on_vaps[BPL_LOAD_STEER_ON_VAPS_LEN])
+{
+
```
remove blank line

adam1985d replied:
was removed in the implementation commit of that function

tomereli commented on framework/platform/bpl/linux/bpl_cfg.cpp:
```diff
@@ -165,6 +165,39 @@ int cfg_get_certification_mode()
+    char ifname[BPL_IFNAME_LEN] = {0};
+    for (int index = 0; index < num_of_interfaces; index++) {
+        if (cfg_get_hostap_iface(index, ifname) == RETURN_ERR) {
+            MAPF_ERR("cfg_get_load_steer_on_vaps: failed to get wifi interface steer vaps for agent"
```
```suggestion
            MAPF_ERR("failed to get wifi interface steer vaps for agent"
```

adam1985d replied:
done

tomereli commented on framework/platform/bpl/uci/cfg/bpl_cfg.cpp:
```diff
@@ -130,6 +130,39 @@ int cfg_get_certification_mode()
                                char load_steer_on_vaps[BPL_LOAD_STEER_ON_VAPS_LEN])
 {
+    if (num_of_interfaces < 1) {
+        MAPF_ERR("cfg_get_load_steer_on_vaps: invalid input: max num_of_interfaces value < 1");
```
```suggestion
        MAPF_ERR("invalid input: max num_of_interfaces value < 1");
```
Same for all other places. Reason - we want to get functions easilly so adding the function name to each print will make grep produce too much output.

adam1985d replied:
was following the syntax used by all current implementations in the bpl_cfg - both for UCI and Linux implementations.
We can add a cleanup task if this is the new approach

adam1985d replied:
fixed my changes as you suggested

tomereli commented on controller/config/beerocks_controller.conf.in:
```diff
@@ -21,7 +21,6 @@ load_client_band_steering=1
 load_client_optimal_path_roaming=1
 load_legacy_client_roaming=1
 load_client_11k_roaming=1
-load_steer_on_vaps=@BEEROCKS_WLAN1_STEERING_VAPS@,@BEEROCKS_WLAN2_STEERING_VAPS@
```
Please squash to the previous commit.

adam1985d replied:
ok - will squash them together

tomereli commented on CMakeLists.txt:
```diff
@@ -126,8 +123,6 @@ set(BEEROCKS_LOG_SYSLOG_ENABLED     "false")
 # Platform specific flags
 if (TARGET_PLATFORM STREQUAL "openwrt")
     if (TARGET_PLATFORM_TYPE STREQUAL "ugw")
-        set(BEEROCKS_WLAN1_STEERING_VAPS "wlan0.1")
```
Same here - squash to a single commit (c40bb11b7ed1f7d477a836de355f4ce8453c59c7)
Commit name can be `tree-wide: remove steer on vaps from conf files`

adam1985d replied:
done

tomereli reviewed (approved):
Minor comments, so approving.

rmelotte commented on framework/platform/bpl/linux/bpl_cfg.cpp:
```diff
@@ -162,6 +162,46 @@ int cfg_get_certification_mode()
 }
 
+int cfg_get_load_steer_on_vaps(int num_of_interfaces,
+                               char load_steer_on_vaps[BPL_LOAD_STEER_ON_VAPS_LEN])
```
Why couldn't load_steer_on_vaps be a `string &` or an std::array?

You would also avoid the need for `load_steer_on_vaps_str`. Or is it intentional and you really want a copy?

adam1985d replied:
it is intentional since the BPL provides a c-like interface and thus no `std::string` or `std::array`.
the upper layer that utilizes the BPL saves the information as a string.

rmelotte reviewed (approved):
Looks fine for me if you had the discussion you mentioned and people agreed.

adam1985d committed:
framework: prplmesh_db: add hostap_iface_steer_vaps to UCI

Preparative Commit.
Currently, the load_steer_on_vaps that is used by the optimal-path-task
is read from the beerocks_controller.conf and is constructed differently
for each platform according to compilation flags.

Added the hostap_iface_steer_vaps param to each radio in the
prplmesh config file. The load_steer_on_vaps will be constructed from
the hostap_iface_steer_vaps of all radios read from UCI.

Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d committed:
framework: bpl: add stubs for cfg_get_load_steer_on_vaps

Preparative Commit.
Added stubs for BPL CFG API to get the load_steer_on_vaps. The
load_steer_on_vaps will be generated from all hostap_iface_steer_vaps
param values configured in UCI for all radios.
The load_steer_on_vaps is used to determine if a VAP is part of a group
of VAPs that the client can be steered between.

Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d committed:
framework: bpl: linux: implement cfg_get_load_steer_on_vaps

Preparative Commit.
Implemented cfg_get_load_steer_on_vaps to construct load_steer_on_vaps
from supported interface names appended with '.0'.
The linux implementation for the 'cfg_get_hostap_iface' returns
hardcoded values for hostap_iface names.
The linux implementation for 'cfg_get_load_steer_on_vaps' also assumes
that the vaps used for steering are the '.0' VAPs.

Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d committed:
framework: bpl: uci: add cfg_get_hostap_iface_steer_vaps

Preparative Commit.
Implemented cfg_get_hostap_iface_steer_vaps - an auxiliary function that
will be called for each radio to get all 'hostap_iface_steer_vaps'
params from UCI and construct the 'load_steer_on_vaps'.

Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d committed:
framework: bpl: uci: implement cfg_get_load_steer_on_vaps

Preparative Commit.
Implemented cfg_get_load_steer_on_vaps to construct load_steer_on_vaps
from 'hostap_iface_steer_vaps' read from UCI for all supported
interfaces.
The API utilizes the 'cfg_get_hostap_iface_steer_vaps' auxiliary function
to read the parameters from UCI.
The resulting 'load_steer_on_vaps' will contain all
'hostap_iface_steer_vaps' concatenated with ',' as delimiter.

Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d committed:
controller: master_main: read load_steer_on_vaps from UCI

set the master configuration of load_steer_on_vaps with value read from
UCI using the 'bpl::cfg_get_load_steer_on_vaps' API.

the 'load_steer_on_vaps', that list the VAPs in groups of VAPs to steer
on, is read by optimal-path-task and used to verify that the current
VAP that the task wants to steer a client from is in the group,
otherwise, the task will abort.

Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d committed:
tree-wide: remove steer on vaps from conf files

the 'load_steer_on_vaps' is read from UCI and is no longer needed in the
beerocks_controller.conf file.
The 'BEEROCKS_WLAN1_STEERING_VAPS', 'BEEROCKS_WLAN2_STEERING_VAPS' and
'BEEROCKS_WLAN3_STEERING_VAPS' compilation flags were used for
platform-specific configuration at compile time of the
'load_steer_on_vaps' in the 'beerocks_controller.conf' file.

removed the 'load_steer_on_vaps' from config_file strunct and from
'read_master_config_file' implementation.
removed 'load_steer_on_vaps' from controller config file.
removed the unneeded compilation flags.

Signed-off-by: Adam Dov <adamx.dov@intel.com>

removed 'load_steer_on_vaps' from controller config file.

Signed-off-by: Adam Dov <adamx.dov@intel.com>

cmake: remove BEEROCKS_WLANx_STEERING_VAPS

the 'BEEROCKS_WLAN1_STEERING_VAPS', 'BEEROCKS_WLAN2_STEERING_VAPS' and
'BEEROCKS_WLAN3_STEERING_VAPS' compilation flags were used for
platform-specific configuration at compile time of the
'load_steer_on_vaps' in the 'beerocks_controller.conf' file.

the 'load_steer_on_vaps' is moved to the UCI and the specified-above
compilation flags can be removed.

removed the unneeded compilation flags.

Signed-off-by: Adam Dov <adamx.dov@intel.com>
